x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -418,34 +418,32 @@ private boolean canAccess(
-                    );
-
-                    // Current user has potentially many roles and therefore potentially many queries
-                    // defining sets of docs accessible
-                    Set<BytesReference> queries = indexAccessControl.getDocumentPermissions().getQueries();
-                    for (BytesReference querySource : queries) {
-                        QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(
-                            querySource,
-                            scriptService,
-                            queryShardContext.getXContentRegistry(),
-                            securityContext.getUser()
-                        );
-                        QueryBuilder rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
-                        if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
-                            // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
-                            return true;
+                if (dls) {
+                    if (licenseChecker.get()) {
+                        // Check to see if any of the roles defined for the current user rewrite to match_all
+
+                        SecurityContext securityContext = new SecurityContext(clusterService.getSettings(), threadContext);
+                        final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+                        final SearchExecutionContext queryShardContext = indexService.newSearchExecutionContext(shardId.id(),
+                            0,
+                            null,
+                            request::nodeStartedTimeMillis,
+                            null,
+                            Collections.emptyMap());
+
+                        // Current user has potentially many roles and therefore potentially many queries
+                        // defining sets of docs accessible
+                        Set<BytesReference> queries = indexAccessControl.getDocumentPermissions().getQueries();
+                        for (BytesReference querySource : queries) {
+                            QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(querySource,
+                                scriptService,
+                                queryShardContext.getXContentRegistry(),
+                                securityContext.getUser());
+                            QueryBuilder rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
+                            if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
+                                // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
+                                return true;
+                            }
                         }
                     }
                     return false;
```

Comment:
This is the only place where we don't return an error but instead just remove the index from the request. Still fail closed though silently. It is possible to also return error here. But it is a bit more involved and I am not sure whether it is worth the effort because:
1. This whole existing method here is a temporary solution. The comment says it should be refactored into (maybe) an interceptor.
2. IIUC, this API is mainly used by Kibana for autocompletion. That means interactive user, not API keys.
3. Silently fail closed still address the security concern which is the main goal.